concrete-test/src/prototyping/prototyper.rs

```diff
@@ -0,0 +1,604 @@
+        &mut self,
+        secret_key: &Self::BinaryLweSecretKeyPrototype,
+        ciphertext_vector: &Self::BinaryLweCiphertextVectorPrototype,
+    ) -> Self::PlaintextVectorPrototype;
+
+    fn decrypt_binary_glwe_ciphertext_to_plaintext_vector(
+        &mut self,
+        secret_key: &Self::BinaryGlweSecretKeyPrototype,
+        ciphertext: &Self::BinaryGlweCiphertextPrototype,
+    ) -> Self::PlaintextVectorPrototype;
+
+    fn decrypt_binary_glwe_ciphertext_vector_to_plaintext_vector(
+        &mut self,
+        secret_key: &Self::BinaryGlweSecretKeyPrototype,
+        ciphertext: &Self::BinaryGlweCiphertextVectorPrototype,
+    ) -> Self::PlaintextVectorPrototype;
+}
+
+impl Prototyper<Precision32> for Maker {
+    type CleartextPrototype = ProtoCleartext32;
+    type CleartextVectorPrototype = ProtoCleartextVector32;
+    type PlaintextPrototype = ProtoPlaintext32;
+    type PlaintextVectorPrototype = ProtoPlaintextVector32;
+    type BinaryLweSecretKeyPrototype = ProtoBinaryLweSecretKey32;
+    type BinaryLweCiphertextPrototype = ProtoBinaryLweCiphertext32;
+    type BinaryLweCiphertextVectorPrototype = ProtoBinaryLweCiphertextVector32;
+    type BinaryGlweSecretKeyPrototype = ProtoBinaryGlweSecretKey32;
+    type BinaryGlweCiphertextPrototype = ProtoBinaryGlweCiphertext32;
+    type BinaryGlweCiphertextVectorPrototype = ProtoBinaryGlweCiphertextVector32;
+    type BinaryBinaryLweKeyswitchKeyPrototype = ProtoBinaryBinaryLweKeyswitchKey32;
+    type BinaryBinaryLweBootstrapKeyPrototype = ProtoBinaryBinaryLweBootstrapKey32;
+
+    fn new_binary_lwe_secret_key(
+        &mut self,
+        lwe_dimension: LweDimension,
+    ) -> Self::BinaryLweSecretKeyPrototype {
+        ProtoBinaryLweSecretKey32(
+            self.core_engine
+                .create_lwe_secret_key(lwe_dimension)
+                .unwrap(),
+        )
+    }
+
+    fn new_binary_glwe_secret_key(
+        &mut self,
+        glwe_dimension: GlweDimension,
+        polynomial_size: PolynomialSize,
+    ) -> Self::BinaryGlweSecretKeyPrototype {
+        ProtoBinaryGlweSecretKey32(
+            self.core_engine
+                .create_glwe_secret_key(glwe_dimension, polynomial_size)
+                .unwrap(),
+        )
+    }
+
+    fn new_binary_binary_lwe_keyswitch_key(
+        &mut self,
+        input_key: &Self::BinaryLweSecretKeyPrototype,
+        output_key: &Self::BinaryLweSecretKeyPrototype,
+        decomposition_level: DecompositionLevelCount,
+        decomposition_base_log: DecompositionBaseLog,
+        noise: Variance,
+    ) -> Self::BinaryBinaryLweKeyswitchKeyPrototype {
+        ProtoBinaryBinaryLweKeyswitchKey32(
+            self.core_engine
+                .create_lwe_keyswitch_key(
+                    &input_key.0,
+                    &output_key.0,
+                    decomposition_level,
+                    decomposition_base_log,
+                    noise,
+                )
+                .unwrap(),
+        )
+    }
```

Comment:
When passing stuff to the core engine in the maker, would it make sense to rely on the sythesizable traits of the prototypes instead of hardcoding the unwrapping (e.g. as in `input_key.0`)? This would have the advantage that if we decide to change the way that prototypes and synthesizing works, we wouldn't have to rewrite all of these functions.

Reply:
Interesting concern and interesting solution ! I did not think about that !

I guess we could do this, but I would actually be worried that it creates more bindings between the prototyping and synthesizing abstractions which are supposed to be two separate things. 

Maybe we could keep it that way for a while, see how it goes ? Anyway its not that much code, so if it turns out to be wrong, it won't be too much work to change it.

Reply:
Sounds reasonable.